src/mlpack/methods/ann/layer/transposed_convolution_impl.hpp

```diff
@@ -0,0 +1,299 @@
+
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename InputDataType,
+    typename OutputDataType
+>
+void TransposedConvolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+>::Reset()
+{
+    weight = arma::cube(weights.memptr(), kW, kH,
+        outSize * inSize, false, false);
+    bias = arma::mat(weights.memptr() + weight.n_elem,
+        outSize, 1, false, false);
+}
+
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename InputDataType,
+    typename OutputDataType
+>
+template<typename eT>
+void TransposedConvolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+>::Forward(const arma::Mat<eT>&& input, arma::Mat<eT>&& output)
+{
+  inputTemp = arma::cube(input.memptr(), inputWidth, inputHeight, inSize);
+  
+  outputWidth = TransposedConvOutSize(inputWidth, kW, dW, padW);
+  outputHeight = TransposedConvOutSize(inputHeight, kH, dH, padH);
+
+  outputTemp = arma::zeros<arma::Cube<eT> >(outputWidth, outputHeight,
+                                            outSize);
+
+  for (size_t outMap = 0, outMapIdx = 0; outMap < outSize; outMap++)
+  {
+    for (size_t inMap = 0; inMap < inSize; inMap++, outMapIdx++)
+    {
+      arma::Mat<eT> convOutput, rotatedFilter;
+      Rotate180(weight.slice(outMapIdx), rotatedFilter);
+
+      BackwardConvolutionRule::Convolution(inputTemp.slice(inMap),
+          rotatedFilter, convOutput, 1, 1);
+      
+      outputTemp.slice(outMap) += convOutput;
+    }
+
+    outputTemp.slice(outMap) += bias(outMap);
+  }
+
+  output = arma::Mat<eT>(outputTemp.memptr(), outputTemp.n_elem, 1);
+}
+
+template<
+    typename ForwardConvolutionRule,
+    typename BackwardConvolutionRule,
+    typename GradientConvolutionRule,
+    typename InputDataType,
+    typename OutputDataType
+>
+template<typename eT>
+void TransposedConvolution<
+    ForwardConvolutionRule,
+    BackwardConvolutionRule,
+    GradientConvolutionRule,
+    InputDataType,
+    OutputDataType
+>::Backward(
+    const arma::Mat<eT>&& /* input */, arma::Mat<eT>&& gy, arma::Mat<eT>&& g)
+{
+  arma::cube mappedError(gy.memptr(), outputWidth, outputHeight, outSize,
+      false, false);
+  std::cout << "mappedError: " << std::endl << mappedError << std::endl;
+  gTemp = arma::zeros<arma::Cube<eT> >(inputTemp.n_rows,
+      inputTemp.n_cols, inputTemp.n_slices);
+  std::cout << "gTemp: " << std::endl << gTemp << std::endl;
+
+  for (size_t outMap = 0, outMapIdx = 0; outMap < outSize; outMap++)
+  {
+    for (size_t inMap = 0; inMap < inSize; inMap++, outMapIdx++)
+    {
+      arma::Mat<eT> output;
+      ForwardConvolutionRule::Convolution(mappedError.slice(outMap),
+          weight.slice(outMapIdx), output, 1, 1);
```

Comment:
@zoq I wasn't able to get `Tensorflow` to provide me with the values for the `Backward` routine. The implementation is most likely correct (Forward is completely debugged and tested), unless we need to rotate the weight matrix here again. Can you think of an alternative way of verifying?

Reply:
I am yet to check for Gradients on all the test cases. Should be done in a few hours :)

Reply:
Sounds good 👍